### PR TITLE
Use boot/ocamlc.opt for building, if available.

### PR DESCRIPTION
**Pull request by stedolan**
With this patch, if the file `boot/ocamlc.opt` exists, it is used for building the compiler instead of `boot/ocamlrun boot/ocamlc`. This means that when working on and repeatedly building the compiler, one can `cp ocamlc.opt boot/` to accelerate builds.

On my machine, this roughly halves the time needed to `make clean; make world` (from 73 to 36 seconds, on a build configured with `--disable-instrumented-runtime --disable-debugger --disable-ocamldoc --disable-graph-lib`).

### CODE REVIEW

**Comment by gasche** [comment]
Gee, I worked on exactly the same thing, and for some reason I never got around to a nice-enough state to send a PR. I like the idea.

I'm not sure if we want to have `boot/ocamlc.opt` in `gitignore`, though. If we have no Makefile targets to handle this file (refresh it, delete it, etc.), and it's not versioned, then maybe I do want to be aware that it is around and it might be stale.

File: Makefile
Comment (gasche) [inline Makefile]: Is this a leftover debug statement, or something you intended to keep in the PR?

Comment (stedolan) [reply Makefile]: Leftover debug! Thanks for spotting.

**Comment by stedolan** [comment]
> I'm not sure if we want to have boot/ocamlc.opt in gitignore, though. If we have no Makefile targets to handle this file (refresh it, delete it, etc.), and it's not versioned, then maybe I do want to be aware that it is around and it might be stale.

I certainly do want it gitignore, as that's the best way to ensure it doesn't get committed. A Makefile target to update/remove/etc. could useful (although I'm not sure that this helps making people aware that it's around). I'm on the fence about whether `make clean` should get rid of it.

**Comment by gasche** [comment]
Makes sense -- those irresponsible `commit -a` users!

So how do we prevent people from shooting themselves in the foot with a stale `boot/ocamlc.opt`?

Maybe we could have warning when `boot/ocamlc.opt` is older than `boot/ocamlc`? In practice this means having to touch it at each checkout, but maybe that is a reasonable habit to take. (I also thought of checking their `-version` output but this is too coarse-grained to be useful.)

**Comment by stedolan** [comment]
I think checking the file dates is a good idea! I'd go further than a warning, though: we should only use `boot/ocamlc.opt` if it is newer than `boot/ocamlc`.

In practice, we won't have to touch at each checkout: when switching branches, git only touches the files that have actually changed. We'll only have to touch `boot/ocamlc.opt` when we switch to a branch that's using a different bootstrap compiler, in which case we probably shouldn't use the old `boot/ocamlc.opt` anyway.

This means that once you `cp ocamlc.opt boot/`, you'll get fast builds until the next update of the bootstrap compiler, after which point you'll have to copy `ocamlc.opt` again.

**Comment by gasche** [comment]
> I think checking the file dates is a good idea! I'd go further than a warning, though: we should only use boot/ocamlc.opt if it is newer than boot/ocamlc.

Okay, but then one issue is silently forgetting to update `ocamlc.opt` -- I'll see in the build logs, though, that `ocamlrun boot/ocamlc` is being used.

> In practice, we won't have to touch at each checkout: when switching branches, git only touches the files that have actually changed.

Nice! I had never checked.

Note: if we had a Makefile target, one nice thing to do would be to write the git commit at the time where ocamlc.opt was generated.

```
boot/ocamlc.opt: ocamlc.opt
    cp ocamlc.opt boot/
    git show --oneline | head -n 1 > boot/ocamlc.opt.stamp
```

(I'm sure people know better git commands that, for example, store the branch name as well.)

**Comment by dra27** [comment]
This seems nice to me, too! I would have `clean` leave `boot/ocamlc.opt` alone but `distclean` should either delete it or display a message that it hasn't cf. https://github.com/ocaml/opam/blob/master/src_ext/Makefile#L227

**Comment by dra27** [comment]
Another thought for promoting its existence - the `opt.opt` could *always* copy `ocamlc.opt` to `boot/ocamlc.opt` if `.git` exists

**Comment by dra27** [comment]
Frankly, it could possibly always copy it - for a release build, it would never be used, as the copy would take place right at the end of `world.opt`

**Comment by gasche** [comment]
I would vote against always promoting `ocamlc.opt` for now. It's a nice idea but it makes everyone, not just expert users, rely on the logic, and suffer from the extra fragility. Let's keep it a you-know-when-you-use-it thing for now, and think about making it the default behavior after we've gathered some experience using it in practice.

I realize now that what resulted in my own experimentation never getting released is probably that I over-thought all these aspects that we just started nitpicking on. @stedolan, there is clearly enough support for your idea to merge (especially if it remains opt-in rather than opt-out). I'll approve now;  please make whatever changes you feel like making inspired by the discussion, and then merge when you are satisfied.

**Review by gasche — APPROVED** [review]

**Comment by stedolan** [comment]
@gasche:
I've changed it to check that ocamlc.opt is newer than ocamlc.

Other Makefile targets, more features, etc. welcome, but I'll leave them out of this PR.

@dra27:
Thanks for pointing out `distclean`. Now, `boot/ocamlc.opt` is deleted by `distclean` but not by `clean`.

I agree with @gasche that we shouldn't promote on every build - most of my compiler builds are broken experiments, and I certainly don't want to use them to build the next compiler!

---

I'm going to wait until Appveyor OK's this (to check that `test -nt` is OK in the windows build), then I'll merge.

**Comment by dra27** [comment]
Oops, in my mind I conflated the promotion of `ocamlrun` during coldstart with bootstrap promotion 😊

**Comment by gasche** [comment]
(@dra27 just fixed a check-typo issue. I happen to be around a bunch of OCaml developers at a big meeting today; could you give me three hours before merging, so that I make sure everyone is appropriately excited by this small-but-noticeable change?)

**Comment by stedolan** [comment]
@dra27 thanks! I'd actually fixed this but forgot to push. Something went wrong with your patch (make complaining about recursive variables?), so I just pushed my version over it. Hopefully check-typo's satisfied now.

@gasche sure, no problem.

**Comment by dra27** [comment]
@stedolan - oops, indeed - I defined `BOOT_OCAMLC` twice where the first should have been `BOOT_OCAML_BYTE`

It's a ~bit~lot nitty, but I kinda preferred my "one-liner" version (albeit with a non-conflicting variable name) to spreading the command over multiple lines?

**Comment by gasche** [comment]
Good news: Damien was fine with the change.

Good bad news: Armaël is asking for a HACKING.adoc explanation of the feature -- and of course he is right.

File: HACKING.adoc
Comment (gasche) [inline HACKING.adoc]: I'm unconvinced by having this in the "intro for beginners" section of HACKING.adoc. Could we move this to the "Development tips and tricks" section, and maybe say a word about the risk of having a stale ocamlc.opt compiler (if you move between branches for example)?

Comment (stedolan) [reply HACKING.adoc]: What's the risk? The makefile checks whether `ocamlc.opt` is newer than `ocamlc` now.

It seemed appropriate to tell people about this feature at the point where we advise them to "keep rebuilding". Why bury the fact that there's a command that makes "keep rebuilding" take half as long?

Comment (dra27) [reply HACKING.adoc]: I think it's correct that the newer test as it stands guards against `ocamlc.opt` being physically unable to build the compiler because of bootstrap issues, but what about if it's compiled with different (experimental) semantics which didn't require a `make promote`/bootstrap?

I agree with @stedolan that this should be here, where it's more likely to be seen and so played with. We can only go so far to make the experience of hacking on the compiler a risk-free endeavour for beginners 🙂

Comment (gasche) [reply HACKING.adoc]: I disagree. Today we *can* give small hacking tasks to almost-complete OCaml beginners, and HACKING.adoc has enough information for them to easily get started building things.

`ocamlc.opt` should not be in the introduction for two reasons:

- There is a risk of things going wrong. Not just a miscompilation due to a stale binary; from a beginner perspective "the build errors/warns because of this out-of-date file" can be disturbing as well.
- This section should be as *simple* as possible, avoiding concept overload. Everything that requires a more complex discussion should be held off for later. If we can remove some information from this section, we should.

For newcomers to compiler hacking, I would rather have slower build time than slower understanding of how to get started on the compiler.

Again, the "Development tips and tricks" section is precisely there to give useful tips to people. For example, `make -C testsuite parallel` halves the testsuite running time, and it's there. The `ocamlc.opt` trick should be there as well.

Comment (stedolan) [reply HACKING.adoc]: I still don't follow point 1 here (out-of-date boot/ocamlc.opt doesn't cause a build error/warning), but point 2 is convincing. I've updated HACKING.adoc.

Comment (gasche) [reply HACKING.adoc]: We're both right for point (1): in my mind there *should* be a warning if `boot/ocamlc.opt` is stale (silently ignoring it is not good UI), but you left it out of the perimeter of the present PR -- I'm fine with that. I hope there will be a warning later.

**Review by gasche — APPROVED** [review]
I like the new documentation, thanks! Good to merge when CI passes.

**Pull request merged**
